File: Cracking the Coding Interview/Trees- Is This a Binary Search Tree/TreesIsThisaBinarySearchTree.java
// https://www.hackerrank.com/challenges/ctci-is-binary-search-tree/problem

/* Hidden stub code will pass a root argument to the function below. Complete the function to solve the challenge. Hint: you may want to write one or more helper functions.  

The Node class is defined as follows:
    class Node {
        int data;
        Node left;
        Node right;
     }
*/
    boolean check(Node root, int min, int max) {
        if (root == null) {
            return true;
        }
        if (root.data <= min || root.data >= max) {
            return false;
        }
        return check(root.left, min, root.data) && check(root.right, root.data, max);
    }

    boolean checkBST(Node root) {
        return check(root, Integer.MIN_VALUE, Integer.MAX_VALUE);
    }

File: Cracking the Coding Interview/Data Structures/Trees- Is This a Binary Search Tree/TreesIsThisaBinarySearchTree.py
# https://www.hackerrank.com/challenges/ctci-is-binary-search-tree/problem

import sys

""" Node is defined as
class node:
    def __init__(self, data):
        self.data = data
        self.left = None
        self.right = None
"""

def check(root, minvalue, maxvalue):
    if root == None:
        return True
    if root.data <= minvalue or root.data >= maxvalue:
        return False
    return check(root.left, minvalue, root.data) and check(root.right, root.data, maxvalue)

def checkBST(root):
    return check(root, -sys.maxsize - 1, sys.maxsize)
